test/Microsoft.AspNetCore.Mvc.Razor.Test/DependencyInjection/MvcRazorMvcBuilderExtensionsTest.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+using Microsoft.AspNetCore.Mvc.Internal;
+using Microsoft.AspNetCore.Mvc.Razor.Internal;
+using Microsoft.AspNetCore.Mvc.Razor.TagHelpers;
+using Microsoft.AspNetCore.Razor.Runtime.TagHelpers;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Test.DependencyInjection
+{
+    public class MvcRazorMvcBuilderExtensionsTest
+    {
+        [Fact]
+        public void AddTagHelpersAsServices_ReplacesTagHelperActivatorAndTagHelperTypeResolver()
+        {
+            // Arrange
+            var services = new ServiceCollection();
+            var builder = services
+                .AddMvc()
+                .ConfigureApplicationPartManager(manager =>
+                {
+                    manager.ApplicationParts.Add(new TestApplicationPart());
+                    manager.FeatureProviders.Add(new TagHelperFeatureProvider());
+                });
+
+            // Act
+            builder.AddTagHelpersAsServices();
+
+            // Assert
+            var activatorDescriptor = Assert.Single(services.ToList(), d => d.ServiceType == typeof(ITagHelperActivator));
+            Assert.Equal(typeof(ServiceBasedTagHelperActivator), activatorDescriptor.ImplementationType);
+
+            var resolverDescriptor = Assert.Single(services.ToList(), d => d.ServiceType == typeof(ITagHelperTypeResolver));
+            Assert.Equal(typeof(FeatureTagHelperTypeResolver), resolverDescriptor.ImplementationType);
+        }
+
+        [Fact]
+        public void AddTagHelpersAsServices_RegistersDiscoveredTagHelpers()
+        {
+            // Arrange
+            var services = new ServiceCollection();
+
+            var manager = new ApplicationPartManager();
+            manager.ApplicationParts.Add(new TestApplicationPart(
+                typeof(TestTagHelperOne),
+                typeof(TestTagHelperTwo)));
+
+            manager.FeatureProviders.Add(new TestFeatureProvider());
+
+            var builder = new MvcBuilder(services, manager);
+
+            // Act
+            builder.AddTagHelpersAsServices();
+
+            // Assert
+            var collection = services.ToList();
```

Comment:
Much nicer if you use `Assert.Collection(services, service => { ...` especially if new services get added or the sequence gets re-ordered.